### ISSUE TITLE
ansible_domain fact can't be parsed using to_nice_yaml filter (seen as object)

### ISSUE THREAD

**Issue by identifystation** [ansible/ansible#20253]
##### ISSUE TYPE
 - Bug Report

##### COMPONENT NAME
ansible_domain fact
to_nice_yaml filter

##### ANSIBLE VERSION
```
v2.2.1.0-0.5.rc5
```

##### OS / ENVIRONMENT
macOS

##### STEPS TO REPRODUCE
```yaml
- name: debug
  debug: var=ansible_domain|to_nice_yaml
```

##### EXPECTED RESULTS
ansible_domain fact to be rendered as YAML

##### ACTUAL RESULTS
Error from to_nice_yaml filter about not being able to represent an object

```
TASK [debug] *******************************************************************
An exception occurred during task execution. To see the full traceback, use -vvv. The error was: RepresenterError: cannot represent an object: vars.local
fatal: [vars]: FAILED! => {"failed": true, "msg": "Unexpected failure during module execution.", "stdout": ""}
```

This is a simple demonstration of the issue. In reality, we template a YAML file by passing in a dictionary defined in our defaults/main.yml and passing it through the to_nice_yml filter. This had been working previously, but is now tripped up by the ansible_domain fact which it appears to view as an object. Attempts to cast it using the jinja2 string filter were ineffective.

This behavior has changed since 2.2.0.0, where it works as expected. We're anticipating the security updates present in RC5 and testing as a result.

**Comment by xenithorb** [comment]
Just seeing this linked to on IRC now, I think I ran into this yesterday with outputting `{{ vars | to_yaml }}` ... it works in most cases but seemed to fail on my more complex plays, I never found out what the exact cause was. It was erroring over an IP address that it thought was an object instead of a string. 

If I can find the time to replicate it again, I'll post here with a similar test-case because I'm confident they're probably related.

**Comment by gilesw** [comment]
I've hit this problem with my custom yaml output filter

https://github.com/gilesw/ansible-filter-plugins

**Comment by sdodson** [comment]
I believe we're seeing the same in 2.2.1.0 when calling a filter.
Example code https://github.com/openshift/openshift-ansible/blob/e810fb6abab0c6fe9198bfc3f39c82ca8054f76e/filter_plugins/openshift_master.py#L471-L488

Traceback
```
    TASK [openshift_master : set_fact] *********************************************
    task path: /home/rdu/sdodson/git/Openshift/openshift-ansible/roles/openshift_master/tasks/main.yml:146
    An exception occurred during task execution. The full traceback is:
    Traceback (most recent call last):
      File "/usr/lib/python2.7/site-packages/ansible/executor/task_executor.py", line 126, in run
        res = self._execute()
      File "/usr/lib/python2.7/site-packages/ansible/executor/task_executor.py", line 443, in _execute
        self._task.post_validate(templar=templar)
      File "/usr/lib/python2.7/site-packages/ansible/playbook/task.py", line 248, in post_validate
        super(Task, self).post_validate(templar)
      File "/usr/lib/python2.7/site-packages/ansible/playbook/base.py", line 373, in post_validate
        value = templar.template(getattr(self, name))
      File "/usr/lib/python2.7/site-packages/ansible/template/__init__.py", line 427, in template
        disable_lookups=disable_lookups,
      File "/usr/lib/python2.7/site-packages/ansible/template/__init__.py", line 383, in template
        disable_lookups=disable_lookups,
      File "/usr/lib/python2.7/site-packages/ansible/template/__init__.py", line 583, in do_template
        res = j2_concat(rf)
      File "<template>", line 9, in root
      File "/home/rdu/sdodson/git/Openshift/openshift-ansible/filter_plugins/openshift_master.py", line 488, in translate_idps
        return yaml.safe_dump([idp.to_dict() for idp in idp_list], default_flow_style=False)
      File "/usr/lib64/python2.7/site-packages/yaml/__init__.py", line 218, in safe_dump
        return dump_all([data], stream, Dumper=SafeDumper, **kwds)
      File "/usr/lib64/python2.7/site-packages/yaml/__init__.py", line 190, in dump_all
        dumper.represent(data)
      File "/usr/lib64/python2.7/site-packages/yaml/representer.py", line 28, in represent
        node = self.represent_data(data)
      File "/usr/lib64/python2.7/site-packages/yaml/representer.py", line 57, in represent_data
        node = self.yaml_representers[data_types[0]](self, data)
      File "/usr/lib64/python2.7/site-packages/yaml/representer.py", line 215, in represent_list
        return self.represent_sequence(u'tag:yaml.org,2002:seq', data)
      File "/usr/lib64/python2.7/site-packages/yaml/representer.py", line 101, in represent_sequence
        node_item = self.represent_data(item)
      File "/usr/lib64/python2.7/site-packages/yaml/representer.py", line 57, in represent_data
        node = self.yaml_representers[data_types[0]](self, data)
      File "/usr/lib64/python2.7/site-packages/yaml/representer.py", line 223, in represent_dict
        return self.represent_mapping(u'tag:yaml.org,2002:map', data)
      File "/usr/lib64/python2.7/site-packages/yaml/representer.py", line 123, in represent_mapping
        node_value = self.represent_data(item_value)
      File "/usr/lib64/python2.7/site-packages/yaml/representer.py", line 67, in represent_data
        node = self.yaml_representers[None](self, data)
      File "/usr/lib64/python2.7/site-packages/yaml/representer.py", line 247, in represent_undefined
        raise RepresenterError("cannot represent an object: %s" % data)
    RepresenterError: cannot represent an object: htpasswd_auth
     
    fatal: [ose3-master.example.com]: FAILED! => {
        "failed": true,
        "msg": "Unexpected failure during module execution.",
        "stdout": ""
    }
```

**Comment by ryansb** [comment]
This looks like it happens when the variable is encoded as a `ansible.vars.unsafe_proxy.AnsibleUnsafeText`, so one fix would be to alter the `AnsibleDumper` class to add a representer for AnsibleUnsafeText (which would likely just be `str(t)` or similar). 

cc/ @bcoca for a second opinion.

I reached this conclusion using the `debug: var=ansible_domain|to_nice_yaml` task and dropping in to find the type that triggers that exception when YAML dumps it.

**Comment by jimi-c** [comment]
Closing This Ticket
===================

Hi!

We believe the above commit should resolve this problem for you. This will also be included in the next release (2.2.2).

If you continue seeing any problems related to this issue, or if you have any further questions, please let us know by stopping by one of the two mailing lists, as appropriate:

   * https://groups.google.com/forum/#!forum/ansible-project - for user questions, tips, and tricks
   * https://groups.google.com/forum/#!forum/ansible-devel - for strategy, future planning, and questions about writing code

Because this project is very active, we're unlikely to see comments made on closed tickets, but the mailing list is a great way to ask questions, or post if you don't think this particular issue is resolved.

Thank you!
